packages/ipfs-message-port-client/src/client.js

```diff
@@ -0,0 +1,348 @@
+      delete queries[id]
+      query.fail(new AbortError())
+      if (this.port) {
+        this.port.postMessage({ type: 'abort', id })
+      }
+    }
+  }
+
+  /**
+   * Sends a given `query` with a given `id` over the message channel.
+   * @param {MessagePort} port
+   * @param {string} id
+   * @param {Query<any, any>} query
+   */
+  static postQuery (port, id, query) {
+    port.postMessage(
+      {
+        type: 'query',
+        namespace: query.namespace,
+        method: query.method,
+        id,
+        input: query.toJSON()
+      },
+      query.transfer()
+    )
+  }
+
+  /**
+   * Handler is invoked when message on the message port is received.
+   * @param {MessageEvent} event
+   */
+  handleEvent (event) {
+    const { id, result } = event.data
+    const query = this.queries[id]
+    // If query with a the given ID is found it is completed with the result,
+    // otherwise it is cancelled.
+    // Note: query may not be found when it was aborted on the client and at the
+    // same time server posted response.
+    if (query) {
+      delete this.queries[id]
+      if (result.ok) {
+        query.succeed(result.value)
+      } else {
+        query.fail(decodeError(result.error))
+      }
+    }
+  }
+}
+exports.Transport = Transport
+
+/**
+ * @template T
+ * @typedef {Array<keyof T>} Keys
+ */
+
+/**
+ * @template T
+ * @typedef {Remote<T> & Service<T>} RemoteService
+ */
+
+/**
+ * Service represents an API to a remote service `T`. It will have all the
+ * methods with the same signatures as `T`.
+ *
+ * @class
+ * @template T
+ */
+class Service {
+  /**
+   * @param {string} namespace - Namespace that remote API is served under.
+   * @param {ProcedureNames<T>} methods - Method names of the remote API.
+   * @param {Transport} transport - Transport to issue queries over.
+   */
+  constructor (namespace, methods, transport) {
+    this.transport = transport
+    // Type script does not like using classes as some dicitionaries, so
+    // we explicitly type it as dictionary.
+    /** @type {Object.<ProcedureNames<T>, Function>} */
+    const api = this
+    for (const method of methods) {
+      /**
+       * @template I, O
+       * @param {I} input
+       * @returns {Promise<O>}
+       */
+      api[method] = input =>
+        this.transport.execute(new Query(namespace, method.toString(), input))
+    }
+  }
+}
+
+/**
+ * Client represents the client to remote `T` service. It is a base clase that
+ * specific API clients will subclass to provide a higher level API for end
+ * user. Client implementations take care of encoding arguments into quries
+ * and issing those to `remote` service.
+ *
+ * @class
+ * @template T
+ */
+class Client {
+  /**
+   * @param {string} namespace
+   * @param {ProcedureNames<T>} methods
+   * @param {Transport} transport
+   */
+  constructor (namespace, methods, transport) {
+    /** @type {RemoteService<T>} */
+    this.remote = (new Service(namespace, methods, transport))
```

Comment:
I'm afraid they are necessary to make TS recognize this as type casting per directive on line 323. I know that's annoying but I think it's still worth it because in return clicking methods on client takes you to implementations on sever, and stuff like that.

Reply:
This example is relatively benign so it's not going to block merging this PR, but please let's not make our code twist itself into weird contortions to satisfy the arbitrary requirements of some external tool.